[ExternalNode] New CI test for VM agent

Anandkumar26 opened the pull request:
Adding a new test for VM/BM agent.
- Test configures namespace, serviceaccount
- Applies vm-agent-rbac.yaml
- Starts antrea service as nodeport
- Verifies create/delete externalentity, validates interface
configuration in OVS

Signed-off-by: Anand Kumar <kumaranand@vmware.com>

XinShuYang commented on ci/jenkins/jobs/projects.yaml:
```diff
@@ -1147,3 +1147,59 @@
+                - ${{sha1}}
+            builders:
+                - builder-e2e-vm
+            trigger_phrase: ^(?!Thanks for your PR).*/test-(e2e-vm).*
```
better to use 'test-vm-e2e' as the trigger phrase.

Anandkumar26 replied:
Sure

XinShuYang commented on ci/jenkins/jobs/projects.yaml:
```diff
@@ -1147,3 +1147,59 @@
              - timeout:
                  timeout: 600
                  type: absolute
+      - '{name}-{test_name}-for-pull-request':
```
Does VMBM need to support networkpolicy and conformance test on new testbed? @wenyingd 
If so I think we need to add them as well.

Anandkumar26 replied:
Later we will be adding networkpolicy test in the VMAgentTest as well, but I think it will be under the same test name. Not sure about conformance

wenyingd replied:
No, I don't think K8s conformance/networkpolicy test are suitable for VM cases. We don't need them

XinShuYang commented on ci/jenkins/jobs/projects.yaml:
```diff
@@ -1147,3 +1147,59 @@
                  timeout: 600
                  type: absolute
+      - '{name}-{test_name}-for-pull-request':
+            test_name: e2e-vm
```
vm-e2e, same for other related items

Anandkumar26 replied:
Sure

XinShuYang commented on ci/jenkins/jobs/macros.yaml:
```diff
@@ -167,3 +167,15 @@
+- builder:
+    name: builder-e2e-vm
+    builders:
+        - shell: |-
```
use consistent indent style

Anandkumar26 replied:
Done

XinShuYang commented on ci/jenkins/jobs/macros.yaml:
```diff
@@ -167,3 +167,15 @@
           exit $((TEST_FAIL_E2E + TEST_FAIL_NP + TEST_FAIL_CONFORMANCE))
+
+- builder:
+    name: builder-e2e-vm
```
builder-vm-e2e

Anandkumar26 replied:
Done

XinShuYang commented on ci/jenkins/jobs/projects.yaml:
```diff
@@ -1147,3 +1147,59 @@
+            branches:
+                - ${{sha1}}
+            builders:
+                - builder-e2e-vm
```
builder-vm-e2e

Anandkumar26 replied:
Done

XinShuYang commented on ci/jenkins/jobs/macros.yaml:
```diff
@@ -167,3 +167,15 @@
+          set -e
+          TEST_FAIL_E2E=0
+          TEST_FAIL_NP=0
+          TEST_FAIL_CONFORMANCE=0
```
TEST_FAIL_E2E=0
TEST_FAIL_NP=0
TEST_FAIL_CONFORMANCE=0
Remove if you don't need them in your script

Anandkumar26 replied:
Sure

XinShuYang commented on ci/jenkins/jobs/projects.yaml:
```diff
@@ -1147,3 +1147,59 @@
+            white_list: '{antrea_white_list}'
+            only_trigger_phrase: true
+            trigger_permit_all: true
+            status_context: jenkins-e2e
```
jenkins-vm-e2e

Anandkumar26 replied:
Done

XinShuYang commented on ci/jenkins/jobs/projects.yaml:
```diff
@@ -1147,3 +1147,59 @@
                  type: absolute
+      - '{name}-{test_name}-for-pull-request':
+            test_name: vm-e2e
+            node: 'antrea-test-node'
```
This is the job label we use. You can change it to antrea-vm-node

Anandkumar26 replied:
Done

XinShuYang commented:
/test-vm-e2e

Anandkumar26 commented:
/test-vm-e2e

Anandkumar26 commented:
/test-vm-e2e

Anandkumar26 commented:
/test-vm-e2e

Anandkumar26 commented:
/test-vm-e2e

Anandkumar26 commented:
/test-vm-e2e

Anandkumar26 commented:
/test-vm-e2e

Anandkumar26 commented:
/test-vm-e2e

Anandkumar26 commented:
/test-vm-e2e

XinShuYang commented:
/test-vm-e2e

Anandkumar26 commented:
/test-vm-e2e

Anandkumar26 commented:
/test-vm-e2e

Anandkumar26 commented:
/test-vm-e2e

Anandkumar26 commented:
/test-vm-e2e

jianjuns commented on docs/vm-installation.md:
```diff
@@ -0,0 +1,78 @@
+Antrea-agent can be run on Linux/Windows VMs. Below are the steps needed to configure and run antrea-agent on VMs
+
+## Prerequisites on cluster
+1. Create a name-space for the antrea agent
```
Namespace

the antrea agent -> `antrea-agent`

Anandkumar26 replied:
Done

jianjuns commented on docs/vm-installation.md:
```diff
@@ -0,0 +1,78 @@
+
+## Prerequisites on cluster
+1. Create a name-space for the antrea agent
+2. Create a service account, cluster role and cluster role bindings for antrea agent.
```
ServiceAccount, ClusterRole, RoleBinding

Anandkumar26 replied:
Done

jianjuns commented on docs/vm-installation.md:
```diff
@@ -0,0 +1,78 @@
+
+Antrea-agent can be run on Linux/Windows VMs. Below are the steps needed to configure and run antrea-agent on VMs
+
+## Prerequisites on cluster
```
To auto-generate pages in antrea.io, we need an empty line after title, bullet list, code/text/yaml blocks.

jianjuns replied:
"on Kubernetes cluster" to be clear.

Anandkumar26 replied:
Done

jianjuns commented on docs/vm-installation.md:
````diff
@@ -0,0 +1,78 @@
+## Prerequisites on cluster
+1. Create a name-space for the antrea agent
+2. Create a service account, cluster role and cluster role bindings for antrea agent.
+```bash
````
Add an empty line before.

jianjuns commented on docs/vm-installation.md:
````diff
@@ -0,0 +1,78 @@
+2. Create a service account, cluster role and cluster role bindings for antrea agent.
+```bash
+kubectl apply -f https://raw.githubusercontent.com/antrea-io/antrea/feature/externalnode/build/yamls/externalnode/vm-agent-rbac.yaml
+```
````
An empty line after.

Anandkumar26 replied:
Done

jianjuns commented on docs/vm-installation.md:
````diff
@@ -0,0 +1,78 @@
+kubectl apply -f https://raw.githubusercontent.com/antrea-io/antrea/feature/externalnode/build/yamls/externalnode/vm-agent-rbac.yaml
+```
+2. Create antrea-agent.kubeconfig file to access K8S API service
+```bash
````
Ditto

Anandkumar26 replied:
Done

jianjuns commented on docs/vm-installation.md:
````diff
@@ -0,0 +1,78 @@
+kubectl config --kubeconfig=antrea-agent.kubeconfig use-context antrea-agent@kubernetes
+```
+3. Create antrea-agent.antrea.kubeconfig file to access antrea controller API service
+```bash
````
Ditto

jianjuns commented on docs/vm-installation.md:
````diff
@@ -0,0 +1,78 @@
+kubectl config --kubeconfig=antrea-agent.antrea.kubeconfig set-context antrea-agent@antrea --cluster=antrea --user=antrea-agent
+kubectl config --kubeconfig=antrea-agent.antrea.kubeconfig use-context antrea-agent@antrea
+```
+4. The antrea-agent configuration file specifies the agent configuration parameters. For all the agent configuration parameters of VM, refer to this [base configuration file](https://raw.githubusercontent.com/antrea-io/antrea/feature/externalnode/build/yamls/externalnode/conf/antrea-agent.conf).
````
Align end of the lines at column 80.

jianjuns replied:
Use relative path to refer to a file in the repo: (../build/yamls/externalnode/conf/antrea-agent.conf)

Anandkumar26 replied:
Done

jianjuns commented on docs/vm-installation.md:
````diff
@@ -0,0 +1,78 @@
+kubectl config --kubeconfig=antrea-agent.antrea.kubeconfig use-context antrea-agent@antrea
+```
+4. The antrea-agent configuration file specifies the agent configuration parameters. For all the agent configuration parameters of VM, refer to this [base configuration file](https://raw.githubusercontent.com/antrea-io/antrea/feature/externalnode/build/yamls/externalnode/conf/antrea-agent.conf).
+Update clientConnection and antreaClientConnection with corresponding kubeconfig file path.
````
We typically quote a component name or a parameter. So, `clientConnection`, `antreaClientConnection`

jianjuns replied:
the corresponding kubeconfig file path

Anandkumar26 replied:
Done

jianjuns commented on docs/vm-installation.md:
```diff
@@ -0,0 +1,78 @@
+Update clientConnection and antreaClientConnection with corresponding kubeconfig file path.
+
+## Prerequisites on Linux VM
+1. OVS needs to be installed. Please refer [Openvswitch](https://github.com/antrea-io/antrea/blob/main/docs/getting-started.md#open-vswitch)
```
refer to

jianjuns replied:
Openvswitch -> the getting-started guide

jianjuns replied:
(getting-started.md#open-vswitch)

jianjuns commented on docs/vm-installation.md:
````diff
@@ -0,0 +1,78 @@
+```
+
+## Prerequisites on Windows VM
+1. OVS needs to be installed. For more information refer [how to install OVS](https://github.com/antrea-io/antrea/blob/main/docs/windows.md#1-optional-install-ovs-provided-by-antrea-or-your-own).
````
"how to install OVS" -> "the Antrea Windows documentation"

jianjuns replied:
(windows.md#1-optional-install-ovs-provided-by-antrea-or-your-own)

Anandkumar26 replied:
Done

wenyingd commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,294 @@
+		// Stop antrea-agent service
+		StopAntreaAgent(t, data, vm)
+		t.Logf("Creating ExternalEntity for %+v", vm)
+		_, err := CreateExternalEntityCrd(data, nsReturned.Name, vm.Nodename, vm.IfName, vm.IP)
```
No need to create ExternalEntity since the Controller logic to convert ExternalNode to ExternalEntity is already added? @mengdie-song What is your point?

Anandkumar26 replied:
Ok. I will update it.

Anandkumar26 commented on docs/vm-installation.md:
````diff
@@ -0,0 +1,78 @@
+kubectl config --kubeconfig=antrea-agent.antrea.kubeconfig set-credentials antrea-agent --token=$TOKEN
+kubectl config --kubeconfig=antrea-agent.antrea.kubeconfig set-context antrea-agent@antrea --cluster=antrea --user=antrea-agent
+kubectl config --kubeconfig=antrea-agent.antrea.kubeconfig use-context antrea-agent@antrea
+```
````
Done

Anandkumar26 commented on docs/vm-installation.md:
```diff
@@ -0,0 +1,78 @@
+4. The antrea-agent configuration file specifies the agent configuration parameters. For all the agent configuration parameters of VM, refer to this [base configuration file](https://raw.githubusercontent.com/antrea-io/antrea/feature/externalnode/build/yamls/externalnode/conf/antrea-agent.conf).
+Update clientConnection and antreaClientConnection with corresponding kubeconfig file path.
+
+## Prerequisites on Linux VM
```
Done

Anandkumar26 commented on docs/vm-installation.md:
````diff
@@ -0,0 +1,78 @@
+kubectl config --kubeconfig=antrea-agent.kubeconfig set-context antrea-agent@kubernetes --cluster=kubernetes --user=antrea-agent
+kubectl config --kubeconfig=antrea-agent.kubeconfig use-context antrea-agent@kubernetes
+```
+3. Create antrea-agent.antrea.kubeconfig file to access antrea controller API service
````
Done

Anandkumar26 commented on docs/vm-installation.md:
```diff
@@ -0,0 +1,78 @@
+# Deploying antrea-agent on VM
+
+Antrea-agent can be run on Linux/Windows VMs. Below are the steps needed to configure and run antrea-agent on VMs
```
Done

Anandkumar26 commented on docs/vm-installation.md:
```diff
@@ -0,0 +1,78 @@
+# Deploying antrea-agent on VM
```
Done

Anandkumar26 reviewed (commented):
Addressed review comments on the installation doc.

Anandkumar26 commented on ci/jenkins/test-vm.sh:
```diff
@@ -0,0 +1,253 @@
+SERVICE_ACCOUNT="vm-agent"
+
+# TODO: change to "control-plane" when testbeds are updated to K8s v1.20
+CONTROL_PLANE_NODE_ROLE="master"
```
I checked the kubernetes version on the master node, it is 1.21. So will remove this comment

Anandkumar26 commented on ci/jenkins/test-vm.sh:
```diff
@@ -0,0 +1,253 @@
+    kubectl apply -f ${WORKDIR}/antrea.yml
+    sleep 30
+    kubectl get nodes -o wide --no-headers=true | awk -v role="$CONTROL_PLANE_NODE_ROLE" '$3 != role {print $6}' | while read IP; do
+        echo "Skip copying new antrea.yml to ${CONTROL_PLANE_NODE_ROLE} node ${IP} home directory"
```
Done

Anandkumar26 commented on ci/jenkins/test-vm.sh:
```diff
@@ -0,0 +1,253 @@
+    echo "Ip address for ubuntu VM ${UBUNTU_VM_IP} and windows VM ${WINDOWS_VM_IP}"
+    kubectl delete sa $SERVICE_ACCOUNT -n $TEST_NAMESPACE --ignore-not-found=true
+    clean_up_one_ns $TEST_NAMESPACE
+    # https://github.com/antrea-io/antrea/issues/1577
```
Done

Anandkumar26 commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,294 @@
+}
+
+func testExternalEntityCommon(t *testing.T, data *TestData, namespace string) {
+	t.Logf("List of Windows VMs: '%s'", testOptions.winVMs)
```
Done

Anandkumar26 commented on ci/jenkins/test-vm.sh:
```diff
@@ -0,0 +1,253 @@
+    echo "Configure antrea-agent as a service on windows VM"
+    # change /tmp/*kubeconfig to C:\antrea-agent\*kubeconfig
+    ssh -o StrictHostKeyChecking=no -i "${WORKDIR}/jenkins_id_rsa" Administrator@${WINDOWS_VM_IP} "sed -i 's/\/tmp\//C:\\\antrea-agent\\\/g' /tmp/antrea-agent.conf"
+    ssh -o StrictHostKeyChecking=no -i "${WORKDIR}/jenkins_id_rsa" Administrator@${WINDOWS_VM_IP} "cp /tmp/antrea-agent.exe C:/antrea-agent/"
```
As an ubuntu user, I cannot scp the files into required directories.  That is the reason, I copy the files into /tmp and then do a sudo copy to respective directories. Keeping it same for windows, for consistency and code readabilitly.

Anandkumar26 commented on ci/jenkins/test-vm.sh:
```diff
@@ -0,0 +1,253 @@
+    kubectl get nodes -o wide --no-headers=true | awk -v role="$CONTROL_PLANE_NODE_ROLE" '$3 != role {print $6}' | while read IP; do
+        echo "Do cleanup on the control plane ${IP}"
+    done
+        CLEAN_LIST=("/tmp/antrea-agent.exe" "/tmp/antctl.exe" "/tmp/antrea-agent.conf" "/tmp/antrea-agent.*kubeconfig" "/tmp/antctl")
```
Sure, will create /tmp/antrea-ci directory and delete the directory as part of cleanup.

Anandkumar26 commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,294 @@
+
+		t.Logf("Creating ExternalNode for %+v", vm)
+		_, err = CreateExternalNodeCrd(data, nsReturned.Name, vm.Nodename, vm.IfName, vm.IP)
+		if err != nil {
```
Working on it, there are multiple statements where I do need to check for err to be nil. Will update them with require.Nul.

Anandkumar26 commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,294 @@
+	for _, vm := range VmList {
+		t.Logf("Verify host interface configuration, VmInfo %+v", vm)
+		exists, err := VerifyInterfaceIsInOVS(t, data, vm)
+		if err != nil {
```
Sure

Anandkumar26 commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,294 @@
+		if err != nil {
+			t.Errorf("ERROR! Failed to verify Host interface, VmInfo %+v", vm)
+		}
+		if exists {
```
Sure

Anandkumar26 commented:
/test-vm-e2e

Anandkumar26 commented:
/test-vm-e2e

Anandkumar26 commented:
/test-vm-e2e

Anandkumar26 commented:
/test-vm-e2e

Anandkumar26 commented:
/test-vm-e2e

Anandkumar26 commented:
/test-vm-e2e

Anandkumar26 commented:
/test-vm-e2e

Anandkumar26 commented on docs/vm-installation.md:
````diff
@@ -0,0 +1,99 @@
+make docker-bin
+```
+
+2. Create antrea-agent service. Below is a sample snippet to start antrea-agent as a service on Ubuntu
````
Done

Anandkumar26 commented on docs/vm-installation.md:
```diff
@@ -0,0 +1,99 @@
+## Prerequisites on Kubernetes cluster
+
+1. Create a Namespace for antrea agent
+2. Create a ServiceAccount, ClusterRole and ClusterRoleBinding for antrea agent.
```
Done

Anandkumar26 commented:
/test-vm-e2e

jianjuns commented on build/yamls/externalnode/vm-agent-rbac.yaml:
```diff
@@ -0,0 +1,88 @@
+kind: ServiceAccount
+metadata:
+  name: vm-agent
+  namespace: test-ns # Change the namespace to where vm-agent is expected to run.
```
Reminder for this one. After more thinking, I feel better to use a more meaningful name like "vm-ns", then we can just use this Namespace in the documentation without a change.

mengdie-song replied:
The rbac file is an old one. Please check my newest change here. https://github.com/antrea-io/antrea/pull/3799/files#diff-3c6589cb5148de813549164a74f4a653cce57c59d20011f69101700d5f8e2d01. We will need ExternalNode privileges added into the yaml.

Anandkumar26 replied:
Yes, the test is using the build/yamls/externalnode/vm-agent-rbac.yml. 
The file build/yamls/externalnode/vm-agent-rbac.yaml is stale now, will delete it.

jianjuns commented on docs/vm-installation.md:
````diff
@@ -0,0 +1,99 @@
+2. Create a ServiceAccount, ClusterRole and ClusterRoleBinding for `antrea-agent`.
+
+```bash
+kubectl apply -f https://raw.githubusercontent.com/antrea-io/antrea/feature/externalnode/build/yamls/externalnode/vm-agent-rbac.yaml
````
We should add the command to create the Namespace?

As commented earlier, I suggest to use "vm-ns" as the default Namespace in the yamls, but we should still add some words to explain you need to update the yamls to use another Namespace.

Anandkumar26 replied:
Sure, I will add command to create NameSpace. We will be changing to 'vm-ns' in build/yamls/externalnode/vm-agent-rbac.yml

mengdie-song commented on docs/vm-installation.md:
```diff
@@ -0,0 +1,99 @@
+## Prerequisites on Kubernetes cluster
+
+1. Create a Namespace for `antrea-agent`
+2. Create a ServiceAccount, ClusterRole and ClusterRoleBinding for `antrea-agent`.
```
Besides these two, maybe we also need to mention that users need to make sure ExternalNode feature switch is enabled for Antrea controller?

Anandkumar26 replied:
Yes, I will add it as a pre-requisite.

jianjuns commented on docs/vm-installation.md:
````diff
@@ -0,0 +1,115 @@
+kubectl create ns vm-ns
+```
+
+3. Create a ServiceAccount, ClusterRole and ClusterRoleBinding for `antrea-agent`
````
Could we also briefly talk about what changes are needed to use a different Namespace other than `vm-ns`?

jianjuns commented on docs/vm-installation.md:
````diff
@@ -0,0 +1,115 @@
+make docker-windows-bin
+```
+
+2. Create `antrea-agent` service using nssm: Below is a sample snippet to start `antrea-agent` as a service
````
Move ":" to the end of the line.

jianjuns commented on docs/vm-installation.md:
```diff
@@ -0,0 +1,115 @@
+
+3. The `antrea-agent` configuration file specifies agent configuration parameters. For all the agent configuration
+   parameters of VM, refer to this [base configuration file](../build/yamls/externalnode/conf/antrea-agent.conf).
+   Update `clientConnection` and `antreaClientConnection` with corresponding the kubeconfig file path.
```
corresponding -> right?

Anandkumar26 replied:
Sure

jianjuns commented on docs/vm-installation.md:
```diff
@@ -0,0 +1,115 @@
+# Installing antrea-agent on VM
+
+`antrea-agent` can be run on Linux/Windows VMs. Below are the steps needed to configure and run
+antrea-agent on VMs
```
Add "." to the end of the sentence.

Anandkumar26 replied:
Done

jianjuns commented on docs/vm-installation.md:
```diff
@@ -0,0 +1,115 @@
+# Installing antrea-agent on VM
+
+`antrea-agent` can be run on Linux/Windows VMs. Below are the steps needed to configure and run
```
How about "This document talks about the steps..."

Anandkumar26 replied:
Changing it to: This document lists/describes the steps needed to configure and run

jianjuns commented on docs/vm-installation.md:
````diff
@@ -0,0 +1,115 @@
+```
+
+3. The `antrea-agent` configuration file specifies agent configuration parameters. For all the agent configuration
+   parameters of VM, refer to this [base configuration file](../build/yamls/externalnode/conf/antrea-agent.conf).
````
"Please refer to this base configuration file to check all configuration parameters for VMs"?

jianjuns reviewed (commented):
A few comments on the document.

jianjuns commented on docs/vm-installation.md:
````diff
@@ -0,0 +1,115 @@
+kubectl config --kubeconfig=antrea-agent.kubeconfig use-context antrea-agent@kubernetes
+```
+
+5. Create antrea-agent.antrea.kubeconfig file to access `antrea-controller` API service
````
Why it is named "antrea-agent.antrea.kubeconfig", but not "antrea-agent.kubeconfig" (I was the latter was the name)?

jianjuns replied:
Add `` around antrea-agent.antrea.kubeconfig: `antrea-agent.antrea.kubeconfig`

jianjuns replied:
Why it is called "API service"? Should we use "API" or "API server"?

Anandkumar26 replied:
Changed it to API server

Anandkumar26 replied:
I think we have references to similar names for running antrea-agent manually for container cases. 
@wenyingd could you please share your thoughts

wenyingd replied:
There are two different kubeconfig files: one is used to connect to APIServer with name `antrea-agent.kubeconfig`, and the other is used to connect to Antrea Controller with name "antrea-agent.antrea.kubeconfig". The middle word "antrea" in second file is actually the cluster name.

For container case, if Antrea Agent is not running inside a Pod, e.g. as a process/Service on Windows, these two kubeconfig files are also needed.

Anandkumar26 commented on docs/vm-installation.md:
```diff
@@ -0,0 +1,115 @@
+
+## Prerequisites on Kubernetes cluster
+
+1. Enable ExternalNode feature on the `antrea-controller`
```
Done

wenyingd commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,281 @@
+	}
+	_, out, _, err := data.provider.RunCommandOnVM(vm.Nodename, cmd)
+	if err != nil {
+		t.Errorf("ERROR! Failed to run command %s on VM %s, err %v", cmd, vm.Nodename, err)
```
No need to output error logs with prefix "ERROR!". Please also modify other error logs in this change.

Anandkumar26 replied:
Sure

wenyingd commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,281 @@
+	}
+
+	if strings.Contains(out, "no row") {
+		t.Errorf("ERROR! Failed to find ovs port %s on VM %s", vm.IfName, vm.Nodename)
```
nit: ovs -> OVS

Anandkumar26 replied:
Done

wenyingd commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,281 @@
+	if vm.OS == "Windows" {
+		cmd = fmt.Sprintf("ovs-vsctl --column=name list port '%s'", vm.IfName)
+	} else {
+		cmd = fmt.Sprintf("sudo ovs-vsctl --column=name list port %s", vm.IfName)
```
Is it working on Linux if we use `'%s'` like the command on Windows?

Anandkumar26 replied:
Yes. the test is passing for both windows and linux

wenyingd commented on ci/jenkins/test-vm.sh:
```diff
@@ -0,0 +1,254 @@
+    echo "====== Applying Antrea yaml ======"
+    ./hack/generate-manifest.sh --enable-nodeport --enable-externalnode > ${WORKDIR}/antrea.yml
+    kubectl apply -f ${WORKDIR}/antrea.yml
+    # TO BE DELETED
```
Remove this?

Anandkumar26 replied:
Done

wenyingd commented on ci/jenkins/test-vm.sh:
```diff
@@ -0,0 +1,254 @@
+
+    echo "Creating files antrea-agent.kubeconfig and antrea-agent.antrea.kubeconfig"
+    # Kubeconfig to access K8S API
+    APISERVER=$(kubectl config view -o jsonpath="{.clusters[?(@.name==\"$CLUSTER_NAME\")].cluster.server}")
```
Is it easier if we provide a script to generate the kubeconfig files, and call it here directly? The script can be used by users too.

Anandkumar26 commented on ci/jenkins/test-vm.sh:
```diff
@@ -0,0 +1,254 @@
+    KUBECONFIG_PATH=${WORKDIR}/.kube/config
+fi
+
+echo "Debug: WORKDIR = $WORKDIR"
```
it is not needed. will remove it.

Anandkumar26 commented on test/e2e/fixtures.go:
```diff
@@ -389,6 +389,58 @@ func exportLogs(tb testing.TB, data *TestData, logsSubDir string, writeNodeLogs
+				// move on to the next VM
+				continue
+			}
+			defer w.Close()
```
I didn't notice it. Actually defer is not needed.  I want to use it only for success case, i.e. when w!=nil.
So probably I will call w.Close() after the w.WriteString() call.

Anandkumar26 commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,281 @@
+	t.Logf("Wait 30s for antrea-controller to convert ExternalNode to ExternalEntity")
+	time.Sleep(30 * time.Second)
+	t.Logf("=== TestVerification ===")
+	for _, vm := range vmList {
```
Sure

Anandkumar26 commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,281 @@
+
+	t.Logf("Wait 60s for antrea-agent to startup and sync externalentities")
+	time.Sleep(60 * time.Second)
+	for _, vm := range vmList {
```
Sure

Anandkumar26 commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,281 @@
+		require.Nil(t, err, "Failed to delete ExternalNode, %+v", err)
+		// Wait for ExternalEntities to be deleted
+		retries := 5
+		for i := 1; i <= retries; i++ {
```
Sure

jianjuns reviewed (commented):
@Anandkumar26 : could you also rebase and resolve conflicts? The PR includes 12 commits and I believe many were merged already.

wenyingd commented on build/charts/antrea/templates/crds/externalnode.yaml:
```diff
@@ -0,0 +1,40 @@
+apiVersion: apiextensions.k8s.io/v1
```
This file is moved to path build/charts/antrea/crds, so this is not needed.

Anandkumar26 replied:
Rebased the patch on top of feature branch. Pls help with review.

wenyingd commented on build/charts/antrea/templates/crds/networkpolicy.yaml:
```diff
@@ -57,6 +57,30 @@ spec:
                     type: object
                     # Ensure that Spec.AppliedTo does not allow NamespaceSelector/IPBlock field
                     properties:
+                      externalEntitySelector:
```
ditto

Anandkumar26 commented on docs/feature-gates.md:
```diff
@@ -323,3 +324,19 @@ Refer to this [document](service-loadbalancer.md) for more information.
+disabled. As of now, this feature requires that `AntreaProxy` and `NetworkPolicyStats` are also enabled.
+
+OVS is required to be installed on the virtual machine or the bare-metal server before running Antrea Agent, and the OVS
+version must be >= 2.13.0 and <= 2.15.1.
```
@wenyingd Could you please respond to this query.

wenyingd commented on docs/vm-installation.md:
````diff
@@ -0,0 +1,114 @@
+1. Generate `antrea-agent` binaries
+
+```powershell
+make docker-windows-bin
````
We should support downloading the binaries from release assets directly.

jianjuns replied:
Ok. We may add downloading after the release.

Anandkumar26 commented on docs/vm-installation.md:
```diff
@@ -0,0 +1,114 @@
+
+## Running antrea-agent on Linux VM
+
+1. Generate `antrea-agent` binaries
```
I have few more changes to be added in installation doc. I will make the change and send it out for review.
I will create an another PR only for documentation, as it can go independently of other patches. Makes it easier to review/update the PR.

jianjuns replied:
A separate PR for documentation makes sense to me.

Anandkumar26 commented on docs/vm-installation.md:
````diff
@@ -0,0 +1,114 @@
+sudo systemctl start antrea-agent
+```
+
+3. The `antrea-agent` configuration file specifies agent configuration parameters. Please refer to this [base configuration file](../build/yamls/externalnode/conf/antrea-agent.conf)
````
Sure, I will update it.

jianjuns commented on hack/generate-manifest.sh:
```diff
@@ -49,6 +49,8 @@ Generate a YAML manifest for Antrea using Helm and print it to stdout.
         --multicast-interfaces        Multicast interface names (default is empty)
         --extra-helm-values-file      Optional extra helm values file to override the default config values
         --extra-helm-values           Optional extra helm values to override the default config values
+        --enable-nodeport             Configure antrea-controller service as nodeport
```
service -> Service
nodeport -> type NodePort

Anandkumar26 replied:
Done

jianjuns commented on hack/generate-manifest.sh:
```diff
@@ -49,6 +49,8 @@ Generate a YAML manifest for Antrea using Helm and print it to stdout.
         --extra-helm-values-file      Optional extra helm values file to override the default config values
         --extra-helm-values           Optional extra helm values to override the default config values
+        --enable-nodeport             Configure antrea-controller service as nodeport
+        --enable-externalnode         Enable externalnode feature in antrea-controller
```
ExternalNode

Anandkumar26 replied:
Done

jianjuns commented on hack/generate-manifest.sh:
```diff
@@ -390,6 +402,14 @@ if $WHEREABOUTS; then
 fi
 
+if $ENABLE_NODEPORT; then
+    HELM_VALUES+=("nodeport.enable=true" "nodeport.port=32767")
```
Better to pass a port number with --enable-nodeport

Anandkumar26 replied:
Sure. I will update it.

jianjuns commented on test/e2e/fixtures.go:
```diff
@@ -389,6 +389,58 @@ func exportLogs(tb testing.TB, data *TestData, logsSubDir string, writeNodeLogs
 		tb.Logf("Error when exporting kubelet logs: %v", err)
 	}
+
+	getVmWriter := func(nodeName, suffix string) *os.File {
```
Vm -> VM

Anandkumar26 replied:
Done

jianjuns commented on test/e2e/fixtures.go:
```diff
@@ -389,6 +389,58 @@ func exportLogs(tb testing.TB, data *TestData, logsSubDir string, writeNodeLogs
+	}
+
+	if testOptions.linVMs != "" {
+		//Export logs for linux VMs.
```
Add a space after //

jianjuns replied:
Linux

Anandkumar26 replied:
Done

jianjuns commented on test/e2e/fixtures.go:
```diff
@@ -389,6 +389,58 @@ func exportLogs(tb testing.TB, data *TestData, logsSubDir string, writeNodeLogs
+	}
+
+	if testOptions.winVMs != "" {
+		//Export logs for windows VMs
```
Add a space after //

Windows

jianjuns commented on test/e2e/utils/enodespecbuilder.go:
```diff
@@ -0,0 +1,28 @@
+package utils
+
+import (
+	crdv1alpha1 "antrea.io/antrea/pkg/apis/crd/v1alpha1"
```
We group imports from different sources and add an empty line between two groups.

antrea.io/antrea imports should be put to the last.

```
import (
        crdv1alpha1 "antrea.io/antrea/pkg/apis/crd/v1alpha1"

	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
)
```

Anandkumar26 replied:
Yes. I'm familiar with it. Missed it in this file. I will update it.

jianjuns commented on test/e2e/utils/enodespecbuilder.go:
```diff
@@ -0,0 +1,28 @@
+package utils
```
The  file should be named: enode_spec_builder.go

jianjuns replied:
ExternalNodeSpecBuilder is quite simple. If it is used only by vmagent_test.go, probably just put the code there. No strong opinion.

Anandkumar26 replied:
The file could be used for other VM tests, so I prefer to keep it in utils.
I followed the same naming conventions as the existing files in test/e2e/utils/.

jianjuns replied:
Unfortunately, they are not named correctly. We use "_" to connect words in file names. You can check other files in the source tree. I would appreciate if you rename other files in test/e2e/utils/ too.

Anandkumar26 replied:
Done

jianjuns commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,276 @@
+const interfaceNameLength = 5
+
+type vmInfo struct {
+	Nodename string
```
NodeName

Anandkumar26 replied:
Done

jianjuns commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,276 @@
+type vmInfo struct {
+	Nodename string
+	OS       string
+	IfName   string
```
In most of Antrea code, we use "IF" not "If".

Anandkumar26 replied:
Sure. I will update it.

Anandkumar26 replied:
Changed this to private, as it not needed outside the package

jianjuns commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,276 @@
+	return false, nil
+}
+
+func createExternalNodeCrd(data *TestData, namespace string, nodename string, ifName string, ip string) (enode *crdv1alpha1.ExternalNode, err error) {
```
Crd -> CRD

Again, all upper cases for abbrs.

Anandkumar26 commented on docs/vm-installation.md:
```diff
@@ -0,0 +1,114 @@
+## Prerequisites on Kubernetes cluster
+
+1. Enable `ExternalNode` feature on the `antrea-controller`
+2. Create a Namespace for `antrea-agent`. This document will use vm-ns as a NameSpace example for illustration.
```
Done

Anandkumar26 commented on docs/vm-installation.md:
````diff
@@ -0,0 +1,114 @@
+kubectl create ns vm-ns
+```
+
+3. Create a ServiceAccount, ClusterRole and ClusterRoleBinding for `antrea-agent`. Update the [RBAC file](../build/yamls/externalnode/vm-agent-rbac.yml)
````
Done

wenyingd commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,276 @@
+				return true, nil
+			} else {
+				t.Logf("ExternalEntity %s is not deleted yet", ee.Name)
+				return false, err
```
Here err is nil, is it what you expected? In this case, the error in 156 is actually because a timeout error.

wenyingd commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,276 @@
+	}
+
+	t.Logf("=== TestVerification ===")
+	for _, vm := range vmList {
```
Use this format for verification?
```

	verifyExternalEntityExistence := func(eeName string, vmNodeName string, expectExists bool) {
		if err := wait.PollImmediate(10*time.Second, 1*time.Minute, func() (done bool, err error) {
			t.Logf("Fetching ExternalEntity %s", eeName)
			_, err = data.crdClient.CrdV1alpha2().ExternalEntities(nsReturned.Name).Get(context.TODO(), eeName, metav1.GetOptions{})
			if err != nil && !errors.IsNotFound(err) {
				t.Logf("Failed to get ExternalEntity %s by ExternalNode %s: %v", eeName, vmNodeName, err)
				return false, err
			}

			if expectExists {
				if err != nil {
					return false, err
				}
				return true, nil
			}
			if err == nil {
				return true, nil
			}
			return false, nil

		}); err != nil {
			op := "created"
			if !expectExists {
				op = "deleted"
			}
			assert.NoError(t, err, "Failed to verify ExternalEntity %s %s by ExternalNode %s", eeName, op, vmNodeName)
		}
	}

	verifyExternalNodeRealization := func(vm vmInfo) {
		if err := wait.PollImmediate(30*time.Second, 1*time.Minute, func() (done bool, err error) {
			t.Logf("Verify host interface configuration, VmInfo %+v", vm)
			exists, err := verifyInterfaceIsInOVS(t, data, vm)
			return exists, err
		}); err != nil {
			assert.NoError(t, err, "Failed to verify host interface in OVS, vmInfo %+v", vm)
			return
		}

		var tempVm vmInfo
		if vm.OS == "Windows" {
			tempVm = getWindowsVmInfo(t, data, vm.Nodename)
		} else {
			tempVm = getVmInfo(t, data, vm.Nodename)
		}
		assert.Exactly(t, vm.IfName, tempVm.IfName, "Failed to verify uplink interface, Expected %s, Got %s", vm.IfName, tempVm.IfName)
		assert.Exactly(t, vm.IP, tempVm.IP, "Failed to verify uplink IP, Expected %s, Got %s", vm.IP, tempVm.IP)
	}

	t.Logf("=== TestVerification ===")
	for _, vm := range vmList {
		hash := sha1.New()
		io.WriteString(hash, vm.IfName)
		hashedIfName := hex.EncodeToString(hash.Sum(nil))
		// Generate ExternalEntity name from ExternalNode
		eeName := vm.Nodename + "-" + hashedIfName[:interfaceNameLength]
		verifyExternalEntityExistence(eeName, vm.Nodename, true)
		verifyExternalNodeRealization(vm)
		err := data.crdClient.CrdV1alpha1().ExternalNodes(nsReturned.Name).Delete(context.TODO(), vm.Nodename, metav1.DeleteOptions{})
		require.Nil(t, err, "Failed to delete ExternalNode")
		verifyExternalEntityExistence(eeName, vm.Nodename, false)
	}
```

Besides, if the test time is long, we could think about use multiple go routines to run verification, each go routine verifies one VM, and we can use a `waitgroup` to ensure all the verify is completed on all VMs.

Anandkumar26 replied:
I don't think the verification part takes time, so the testrun would finish quickly.

Anandkumar26 replied:
Thanks for the suggestion, this looks much cleaner

Anandkumar26 commented on test/e2e/fixtures.go:
```diff
@@ -389,6 +389,58 @@ func exportLogs(tb testing.TB, data *TestData, logsSubDir string, writeNodeLogs
 		tb.Logf("Error when exporting kubelet logs: %v", err)
 	}
+
+	getVMWriter := func(nodeName, suffix string) *os.File {
```
Makes sense. Will update it.

Anandkumar26 commented on test/e2e/fixtures.go:
```diff
@@ -389,6 +389,58 @@ func exportLogs(tb testing.TB, data *TestData, logsSubDir string, writeNodeLogs
+		return f
+	}
+
+	if testOptions.linVMs != "" {
```
Done. Thanks

Anandkumar26 commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,276 @@
+	IFIndex  string // Used only for Windows
+}
+
+// TestBasic is the top-level test which contains some subtests for
```
Done

Anandkumar26 commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,276 @@
+		assert.Exactly(t, vm.IP, tempVm.IP, "Failed to verify uplink IP, Expected %s, Got %s", vm.IP, tempVm.IP)
+	}
+
+	t.Logf("=== TestCleanUp ===")
```
Done

Anandkumar26 commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,276 @@
+
+	t.Logf("Wait 30s for antrea-agent to sync externalentities")
+	time.Sleep(30 * time.Second)
+	t.Logf("=== Verification after TestCleanUp ===")
```
Done

Anandkumar26 commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,276 @@
+			if err != nil {
+				t.Logf("Could not get ExternalEntity %s by ExternalNode %s", eeName, vm.NodeName)
+				return false, err
+			} else {
```
Done

Anandkumar26 commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,276 @@
+
+	for _, vm := range vmList {
+		// Verify ExternalEntity is processed by antrea-agent
+		if err := wait.PollImmediate(30*time.Second, 1*time.Minute, func() (done bool, err error) {
```
We can increase based on the testing results.

Anandkumar26 replied:
As we are processing Linux VMs first, it should be okay.

Anandkumar26 commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,276 @@
+		// Generate ExternalEntity name from ExternalNode
+		eeName := vm.NodeName + "-" + hashedIfName[:interfaceNameLength]
+		// Verify ExternalEntities are created
+		if err := wait.PollImmediate(10*time.Second, 1*time.Minute, func() (done bool, err error) {
```
We are processing one VM at a time.

Anandkumar26 commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,276 @@
+// TestBasic is the top-level test which contains some subtests for
+// basic test cases so they can share setup, teardown.
+func TestVMAgent(t *testing.T) {
+	skipIfHasWindowsNodes(t)
```
let me check and add it.

jianjuns reviewed (commented):
BTW, I would change "CI/CD" to "CI" in the commit message. We do not have CD for Antrea.

wenyingd commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,285 @@
+	for _, vm := range vmList {
+		stopAntreaAgent(t, data, vm)
+		t.Logf("Creating ExternalNode for %+v", vm)
+		_, err := createExternalNodeCRD(data, nsReturned.Name, vm.nodeName, vm.ifName, vm.ip)
```
Use `namespace` directly?

Anandkumar26 replied:
Sure, for now will add a global variable for nameSpace and serviceAccount

wenyingd commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,285 @@
+	verifyExternalEntityExistence := func(eeName string, vmNodeName string, expectExists bool) {
+		if err := wait.PollImmediate(10*time.Second, 1*time.Minute, func() (done bool, err error) {
+			t.Logf("Verifying ExternalEntity %s, expectExists %t", eeName, expectExists)
+			_, err = data.crdClient.CrdV1alpha2().ExternalEntities(nsReturned.Name).Get(context.TODO(), eeName, metav1.GetOptions{})
```
ditto

wenyingd commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,285 @@
+		eeName := vm.nodeName + "-" + hashedIfName[:interfaceNameLength]
+		verifyExternalEntityExistence(eeName, vm.nodeName, true)
+		verifyExternalNodeRealization(vm)
+		err := data.crdClient.CrdV1alpha1().ExternalNodes(nsReturned.Name).Delete(context.TODO(), vm.nodeName, metav1.DeleteOptions{})
```
ditto

wenyingd commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,285 @@
+}
+
+func testExternalNodeBasic(t *testing.T, data *TestData, namespace string) {
+	t.Logf("List of Windows VMs: '%s', Linux VMs: '%s'", testOptions.winVMs, testOptions.linVMs)
```
Could you create a unique function to prepare setup, so that we could use the function in other test at the beginning?

wenyingd replied:
Think through, the creation for serviceaccount, namespace, ExternalNode, and start antrea-agent should be out of the basic test. And we should also have a defer action out basic test to delete ExternalNode/serviceaccount/namespace after all test cases are done.

wenyingd commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,285 @@
+	}
+
+	t.Logf("TestSetup")
+	for _, vm := range vmList {
```
Also provide a function for ExternalNode creation and start antrea-agent on VM?

Anandkumar26 replied:
Sure

Anandkumar26 commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,285 @@
+			}
+			// ExternalEntity is expected to be deleted, check for error
+			if err != nil {
+				if errors.IsNotFound(err) {
```
Done

wenyingd commented on docs/vm-installation.md:
````diff
@@ -0,0 +1,114 @@
+1. Generate `antrea-agent` binaries
+
+```bash
+make docker-bin
````
No, the binaries are inlucded in Antrea releases.

wenyingd commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,307 @@
+func testExternalNode(t *testing.T, data *TestData, vmList []vmInfo) {
+	verifyExternalNodeRealization := func(vm vmInfo) {
+		if err := wait.PollImmediate(30*time.Second, 1*time.Minute, func() (done bool, err error) {
+			t.Logf("Verify host interface configuration, VmInfo %+v", vm)
```
ditto

Anandkumar26 replied:
Here it's dumping vmInfo object, because we are verifying Interfaces are present in OVS
The logs will contain the interface name that is to be present, which would help in troubleshooting.

wenyingd reviewed (commented):
Two major points: 1) do not expose VM IP in the logs even for debugging. This is because the VMs IP are real IP in the private lab, and we should not publish it. 2) do not rename other files if they are not related with this change.

Anandkumar26 commented on test/e2e/providers/remote.go:
```diff
@@ -68,6 +68,15 @@ func (p *RemoteProvider) RunCommandOnNodeExt(nodeName, cmd string, envs map[stri
 	return exec.RunSSHCommand(host, clientCfg, cmd, envs, stdin, sudo)
 }
 
+func (p *RemoteProvider) RunCommandOnVM(nodeName, cmd string) (
```
This was done to identify and differentiate K8s node vs VMs, so that any test consumer who wants to run commands on Nodes and VMs can make separate calls.

Anandkumar26 replied:
Discussed offline and we decided to reuse `RunCommandOnNode ` function for VM case as well.

Anandkumar26 commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,307 @@
+// Copyright 2022 Antrea Authors
```
Sure make sense. I will take this for any future changes.

Anandkumar26 commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,307 @@
+	if testOptions.linVMs != "" {
+		vms := strings.Split(testOptions.linVMs, ",")
+		for _, vm := range vms {
+			t.Logf("Get info for Linux VM: %s", vm)
```
Here we are just dumping VM node name. 
    
    sample output from logs:
    
    vmagent_test.go:83: List of Windows VMs: 'vmbmtest0-win-0', Linux VMs: 'vmbmtest0-1'
    vmagent_test.go:86: Using service account vm-agent
    vmagent_test.go:87: Using namespace vm-ns
    vmagent_test.go:92: Get info for Linux VM: vmbmtest0-1
    vmagent_test.go:100: Get info for Windows VM, vmbmtest0-win-0

Anandkumar26 commented on ci/jenkins/test-vm.sh:
```diff
@@ -0,0 +1,249 @@
+        fi
+    done
+
+    echo "IP address of Linux VM ${UBUNTU_VM_IP} and Windows VM ${WINDOWS_VM_IP}"
```
The IP addresses are internal IPs. They cannot be accessed outside VMware network.
But I can remove logs which dumps Ip addresses.

Anandkumar26 commented on ci/jenkins/test-vm.sh:
```diff
@@ -0,0 +1,249 @@
+    echo "IP address of Linux VM ${UBUNTU_VM_IP} and Windows VM ${WINDOWS_VM_IP}"
+    kubectl delete sa $SERVICE_ACCOUNT -n $TEST_NAMESPACE --ignore-not-found=true
+    clean_up_one_ns $TEST_NAMESPACE
+    echo "Deleting stale antrea-agent files from ${UBUNTU_VM_IP} and ${WINDOWS_VM_IP}"
```
Sure

Anandkumar26 commented on ci/jenkins/test-vm.sh:
```diff
@@ -0,0 +1,249 @@
+    echo "Updating antrea-agent.conf"
+    sed -i 's|#externalNodeNamespace: default|externalNodeNamespace: vm-ns|g' ${WORKDIR}/antrea-agent.conf
+    sed -i 's|kubeconfig: |kubeconfig: /tmp/antrea-ci/|g' ${WORKDIR}/antrea-agent.conf
+    echo "Copying binaries and conf to Linux VM: $UBUNTU_VM_IP"
```
Sure

Anandkumar26 commented on ci/jenkins/test-vm.sh:
```diff
@@ -0,0 +1,249 @@
+    scp -q -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -i "${WORKDIR}/jenkins_id_rsa" ./bin/antrea-agent ubuntu@${UBUNTU_VM_IP}:/tmp/antrea-ci/antrea-agent
+    scp -q -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -i "${WORKDIR}/jenkins_id_rsa" ./bin/antctl ubuntu@${UBUNTU_VM_IP}:/tmp/antrea-ci/antctl
+    scp -q -o StrictHostKeyChecking=no -i "${WORKDIR}/jenkins_id_rsa" ${WORKDIR}/antrea-agent.conf  ubuntu@${UBUNTU_VM_IP}:/tmp/antrea-ci/antrea-agent.conf
+    echo "Copying binaries and conf to Windows VM: $WINDOWS_VM_IP"
```
Sure

wenyingd reviewed (approved):
The code change for ExternalNode e2e test looks good to me, but we should wait for its dependent path https://github.com/antrea-io/antrea/pull/3799 merged first.

antoninbas commented on build/charts/antrea/values.yaml:
```diff
@@ -298,3 +298,7 @@ testing:
   simulator:
     enable: false
+
+nodeport:
+  enable: false
+  port: 32767
```
* there is no documentation for this
* the name `nodeport` is to vague
* can we use absence of value for `port` / a zero value as meaning "don't enable NodePort" instead of having an extra boolean flag?
* this should be under `controller:` IMO

wenyingd replied:
The documentation on using NodePort for antrea-controller will be added in #3963

Anandkumar26 replied:
Updated the PR to remove to consider zero value as disabling NodePort.
For the documentation, we have leveraged the existing framework to ssh into the VM (i.e. external node). So I don't think we need a separate/documentation for the VM test.

antoninbas replied:
> For the documentation, we have leveraged the existing framework to ssh into the VM (i.e. external node). So I don't think we need a separate/documentation for the VM test.

I don't understand this comment. I was pointing to the lack of comment to document this Helm value.

Anandkumar26 replied:
Updated it.

antoninbas commented on build/yamls/externalnode/conf/antrea-agent.conf:
```diff
@@ -32,6 +32,18 @@ featureGates:
+  # The policyBypassRules describes the traffic that is expected to bypass NetworkPolicy rules.
+  # Each rule contains the following four attributes:
+  # direction (ingress|egress), protocol(tcp/udp/icmp/ip), remote IP or CIDR, dst port (ICMP doesn't require),
+  # and these attributes are separated by ":", e.g. ["ingress:tcp:1.1.1.1:22", "egress:tcp:2.2.2.2:80"].
```
this is a YAML file, so why don't we use more structured data?
```yaml
policyBypassRules:
    - direction: ingress
      protocol: tcp
      cidr: 1.1.1.1/32
      port: 22
```

wenyingd replied:
This code is from this PR https://github.com/antrea-io/antrea/pull/3799 @mengdie-song would you address it or share your thoughts?

mengdie-song replied:
Thanks for the suggestion. I will update this part in my change.

antoninbas commented on build/yamls/externalnode/vm-agent-rbac.yml:
```diff
@@ -0,0 +1,92 @@
+  - kind: ServiceAccount
+    name: vm-agent
+    namespace: vm-ns # Change the Namespace to where vm-agent is expected to run.
+
```
empty line

antoninbas commented on ci/jenkins/jobs/macros.yaml:
```diff
@@ -157,3 +157,12 @@
+          #!/bin/bash
+          set -e
+          DOCKER_REGISTRY="$(head -n1 ci/docker-registry)"
+          ./ci/jenkins/test-vm.sh  --registry ${DOCKER_REGISTRY} --kubeconfig /var/lib/jenkins/.kube/config
```
missing newline character at the end of file

Anandkumar26 replied:
Done

antoninbas commented on hack/generate-manifest.sh:
```diff
@@ -84,6 +86,9 @@ MULTICAST=false
 MULTICAST_INTERFACES=""
 HELM_VALUES_FILES=()
 HELM_VALUES=()
+ENABLE_NODEPORT=false
+NODEPORT_NUMBER=32767
```
names are too generic

Anandkumar26 replied:
Updated it. is this better?
EXPORT_CONTROLLER_SERVICE=false
NODEPORT=32767

antoninbas replied:
I think you only need one variable (`ANTREA_SERVICE_NODEPORT`) with 0 meaning "NodePort disabled"

Anandkumar26 replied:
I have updated it.

antoninbas commented on cmd/antrea-agent/agent.go:
```diff
@@ -328,7 +332,16 @@ func run(o *Options) error {
 	// notifying NetworkPolicyController and EgressController to reconcile rules
 	// related to the updated Pods.
-	podUpdateChannel := channel.NewSubscribableChannel("PodUpdate", 100)
+	var podUpdateChannel *channel.SubscribableChannel
+	// externalEntityUpdateChannel is a channel for receiving ExternalEntity updates from ExternalNodeController and
+	// notifying NetworkPolicyController to reconcile rules related to the updated ExternalEntities.
+	var externalEntityUpdateChannel *channel.SubscribableChannel
+	if o.nodeType == config.K8sNode {
+		podUpdateChannel = channel.NewSubscribableChannel("PodUpdate", 100)
+	} else {
+		externalEntityUpdateChannel = channel.NewSubscribableChannel("ExternalNodeUpdate", 100)
+	}
```
why don't we use a single channel to keep things simpler? only one of them is ever used anyway and the channel is not strongly typed.

wenyingd replied:
@mengdie-song

mengdie-song replied:
Thanks for the comment. It is true that only one of them will be used at one time, but I am thinking that it is much clearer to have separate channels for Pod and ExternalEntity. 
In this way, it is explicit to show that ExternalNodeController consumes ExternalEntityUpdate channel and others(e.g. NetworkPolicyController, EgressController, cniServer...) consumes PodUpdateChannel. And we will keep the behavior of having only one type of object in one specific channel and do not need to make a name like "EntityUpdate" for a mixed channel. Does it make sense to you? @antoninbas 

The comment reminds me that the name here should be ExternalEntityUpdate instead of ExternalNodeUpdate, will update it.

antoninbas replied:
I was suggesting:
```go
var localEndpointUpdateChannel *channel.SubscribableChannel
if o.nodeType == config.K8sNode {
        localEndpointUpdateChannel = channel.NewSubscribableChannel("PodUpdate", 100)
} else {
        localEndpointUpdateChannel = channel.NewSubscribableChannel("ExternalNodeUpdate", 100)
}
```
so the channels would still have different names

I don't feel strongly about this though, so if @tnqn and @jianjuns have already reveiwed and approved that part, feel free to ignore my comment.

mengdie-song replied:
Thanks Antonin. I prefer to keep two channels internally for now and we can enhance this part later if folks do not like it.

antoninbas commented on pkg/agent/util/net.go:
```diff
@@ -415,3 +415,17 @@ func GenerateRandomMAC() net.HardwareAddr {
 	return buf
 }
+
+func GetGlobalIPNetsByName(link *net.Interface) ([]*net.IPNet, error) {
```
It feels like the name of this function is not accurate since there is no `name` parameter...

antoninbas commented on pkg/agent/util/net_linux.go:
```diff
@@ -240,6 +241,34 @@ func DeleteOVSPort(brName, portName string) error {
+	if _, ok := err.(netlink.LinkNotFoundError); ok {
+		return false
+	}
+	klog.ErrorS(err, "Failed to find host interface", "name", ifName)
```
why log an error? that doesn't make sense for this function. If the caller needs the interface to exist, then the caller should be the one logging the error when it is missing

Anandkumar26 replied:
@mengdie-song 
Can you please take care of this in your patch?

antoninbas commented on pkg/agent/util/net_windows.go:
```diff
@@ -895,3 +895,13 @@ func ReplaceNetNeighbor(neighbor *Neighbor) error {
+
+// TODO: Implement GetInterfaceConfig for Windows
+func GetInterfaceConfig(ifName string) (*net.Interface, []*net.IPNet, []interface{}, error) {
+	return nil, nil, nil, nil
+}
+
+// TODO: Implement RenameInterface for Windows
+func RenameInterface(from, to string) error {
+	return nil
```
we should probably return an error for now?

Anandkumar26 replied:
I have a windows patch which implements this function, so this should be okay.

antoninbas commented on test/e2e/fixtures.go:
```diff
@@ -389,6 +389,32 @@ func exportLogs(tb testing.TB, data *TestData, logsSubDir string, writeNodeLogs
+			w.Close()
+		}
+	}
+	if testOptions.linVMs != "" {
+		cmd := "sudo cat /var/log/antrea/antrea-agent.log"
+		writeVMAgentLog(cmd, testOptions.linVMs)
+	}
+	if testOptions.winVMs != "" {
```
spell out linux and windows

Anandkumar26 replied:
Sure

antoninbas commented on test/e2e/fixtures.go:
```diff
@@ -389,6 +389,32 @@ func exportLogs(tb testing.TB, data *TestData, logsSubDir string, writeNodeLogs
+		}
+	}
+	if testOptions.linVMs != "" {
+		cmd := "sudo cat /var/log/antrea/antrea-agent.log"
```
I'd rather not have the command use `sudo`. Notice that the `journalctl` command above doesn't use sudo for example. Maybe we can assume that the user we ssh as, has permissions to read the file? Is that not the case for our testbed?

antoninbas commented on test/e2e/utils/enode_spec_builder.go:
```diff
@@ -0,0 +1,43 @@
+// Copyright 2022 Antrea Authors.
```
spell our externalnode in the file name

Anandkumar26 replied:
Sure

antoninbas commented on test/e2e/main_test.go:
```diff
@@ -84,6 +84,8 @@ func testMain(m *testing.M) int {
 	flag.BoolVar(&testOptions.flowVisibility, "flow-visibility", false, "Run flow visibility tests")
 	flag.StringVar(&testOptions.coverageDir, "coverage-dir", "", "Directory for coverage data files")
 	flag.StringVar(&testOptions.skipCases, "skip", "", "Key words to skip cases")
+	flag.StringVar(&testOptions.linVMs, "linVMs", "", "hostname of Linux VMs")
+	flag.StringVar(&testOptions.winVMs, "winVMs", "", "hostname of Windows VMs")
```
spell out linux and windows

Anandkumar26 replied:
Sure

antoninbas commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,307 @@
+// TestVMAgent is the top-level test which can contain some subtests for
+// VMAgent so they can share setup, teardown.
+func TestVMAgent(t *testing.T) {
+	skipIfFeatureDisabled(t, features.ExternalNode, false, true)
```
I feel like this is not sufficient. What if I enable the feature gate (or maybe the feature gate graduates to beta, so it's enabled by default), but I want to run the "normal" tests, and I don't provide a list of VMs? Shouldn't the tests be skipped in this case?

Anandkumar26 replied:
This test is specific to VM use-case and it is not testing anything on the cluster. 
Do you have any recommendations on how we could disable/skips the tests if its not required

antoninbas replied:
Why not skip if the VM list is empty?

Anandkumar26 replied:
Sure. I will add the required change.

antoninbas commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,307 @@
+		// Generate ExternalEntity name from ExternalNode
+		eeName := vm.nodeName + "-" + hashedIfName[:interfaceNameLength]
+		err := data.crdClient.CrdV1alpha1().ExternalNodes(nameSpace).Delete(context.TODO(), vm.nodeName, metav1.DeleteOptions{})
+		require.Nil(t, err, "Failed to delete ExternalNode")
```
that probably shouldn't be a `require` assertion, we want to keep on cleaning even if one deletion fails

Anandkumar26 replied:
Done. I will just dump a log message here.

antoninbas commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,307 @@
+			} else {
+				tempVm = getWindowsVMInfo(t, data, vm.nodeName)
+			}
+			if vm.ifName != tempVm.ifName {
+				t.Errorf("Failed to verify uplink interface, Expected %s, Got %s", vm.ifName, tempVm.ifName)
+				return false, nil
+			}
+			if vm.ip != tempVm.ip {
+				t.Errorf("Failed to verify uplink IP, Expected %s, Got %s", vm.ip, tempVm.ip)
+				return false, nil
+			}
```
sometimes you use testify assertions, and sometimes you use an if statement and `t.Error`. Why not be consistent? You can use `assert.Equal` here for example

Anandkumar26 replied:
These checks are inside a wait.PollImmediate(), which we expect to return true within a specified timeout. 
That is the reason, we are not have ing assert check in this block.

antoninbas replied:
I don't understand your comment. There is no functional difference between using `t.Error` and `assert`. This code is equivalent to:
```go
			if !assert.Equal(t, vm.ifName, tempVm.ifName, "Unexpected uplink interface name") {
				return false, nil
			}
			if !assert.Equal(t, vm.ip, tempVm.ip, ""Unexpected uplink interface IP") {
				return false, nil
			}
```

Anandkumar26 replied:
I did not know both are same. I will incorporate your suggestion.

antoninbas commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,307 @@
+	} else {
+		cmd = "sudo systemctl start antrea-agent"
+	}
+	_, _, _, err := data.RunCommandOnNode(vm.nodeName, cmd)
```
I noticed that most of the time you ignore stderr / stdout. Are these not useful for debugging in case of failure?

Anandkumar26 replied:
The commands are the basic functions to start and stop antrea-agent. We do capture stdoout, whenever we need an output. for the failure cases, we do capture the err returned from the command, so I'm not sure if we need stderr output to be logged.

antoninbas replied:
> The commands are the basic functions to start and stop antrea-agent

Things can still go wrong.

If I am not mistaken, you use the remove provider and this will execute the command over SSH. `err` will only be `nil` if we fail to establish an SSH session. If the command run over SSH fails (i.e., returns a non-zero exit code), `err` will be `nil`. https://github.com/antrea-io/antrea/blob/main/test/e2e/providers/exec/ssh.go#L62-L64. That's why it's important to look at the first return value (exit code) and log stderr in case of a non-zero exit code.

antoninbas commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,307 @@
+func startAntreaAgent(t *testing.T, data *TestData, vm vmInfo) {
+	t.Logf("Starting antrea-agent on VM: %s", vm.nodeName)
+	var cmd string
+	if vm.osType == "Windows" {
+		cmd = "nssm start antrea-agent"
+	} else {
+		cmd = "sudo systemctl start antrea-agent"
+	}
```
wouldn't it be easier to run it as a process for the duration of the test?

Anandkumar26 replied:
From a user point of view, running it as a service is preferred. So that is the reason, we decided to go the service way for the test as well.

antoninbas commented on build/charts/antrea/values.yaml:
```diff
@@ -248,6 +248,9 @@ controller:
     resources:
       requests:
         cpu: "200m"
+  # Exporting antrea controller service via nodeport
+  nodeport:
+    port: 0
```
this should be moved up next to `apiPort` for clarity.

I don't understand the nexting. Why not just:
```
apiNodePort: 0
```

Anandkumar26 replied:
Updated it.

antoninbas commented on build/charts/antrea/values.yaml:
```diff
@@ -207,6 +207,8 @@ agent:
   # -- Port for the antrea-controller APIServer to serve on.
   apiPort: 10349
+  # -- Exporting antrea-controller APIServer via nodeport.
+  nodeport: 0
```
this should be `nodePort` for consistency with other params and the rest of the code base. I actually think we should use `apiNodePort` to make it more consistent with `apiPort`.

Anandkumar26 replied:
Updated it.

antoninbas commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,306 @@
+		hashedIfName := hex.EncodeToString(hash.Sum(nil))
+		// Generate ExternalEntity name from ExternalNode
+		eeName := vm.nodeName + "-" + hashedIfName[:interfaceNameLength]
+		t.Logf("Running verifyExternalEntityExistence")
+		verifyExternalEntityExistence(t, data, eeName, vm.nodeName, true)
+		t.Logf("Running verifyExternalNodeRealization")
+		verifyExternalNodeRealization(vm)
```
why did you use a closure for `verifyExternalNodeRealization` and a regular function for `verifyExternalEntityExistence`?

Anandkumar26 replied:
verifyExternalEntityExistence - Is used by teardownVMAgentTest() and the actual test(testExternalNode()) to verify if an ExternalEntity is added/deleted. That is the reason, its a regular function.

antoninbas commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,306 @@
+		assert.Exactly(t, vm.ip, tempVm.ip, "Failed to verify uplink IP, Expected %s, Got %s", vm.ip, tempVm.ip)
+	}
+	for _, vm := range vmList {
+		hash := sha1.New() // #nosec G401: not used for security purposes
+		io.WriteString(hash, vm.ifName)
+		hashedIfName := hex.EncodeToString(hash.Sum(nil))
+		// Generate ExternalEntity name from ExternalNode
+		eeName := vm.nodeName + "-" + hashedIfName[:interfaceNameLength]
```
make a helper function for this?

Anandkumar26 replied:
Done

antoninbas commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,306 @@
+	cmd = "ip -o -4 route show to default | awk '{print $5}'"
+	vm.osType = "Linux"
+	rc, ifName, _, err := data.RunCommandOnNode(nodeName, cmd)
+	require.Nil(t, err, "Failed to run command %s on VM %s, err %v, rc %d", cmd, nodeName, err, rc)
```
use `require.NoError` for errors, not `require.Nil`. It makes the intent more explicit and in case of an error, the test output is more helpful

Anandkumar26 replied:
Sure. Updated it.

antoninbas commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,306 @@
+	cmd = "ip -o -4 route show to default | awk '{print $5}'"
+	vm.osType = "Linux"
+	rc, ifName, _, err := data.RunCommandOnNode(nodeName, cmd)
+	require.Nil(t, err, "Failed to run command %s on VM %s, err %v, rc %d", cmd, nodeName, err, rc)
```
should check for `rc == 0` as well, same everywhere else

Anandkumar26 replied:
Done.

tnqn commented:
this PR needs rebasing

Anandkumar26 commented:
> this PR needs rebasing

Rebased on top of feature/externalnode branch.

antoninbas commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,327 @@
+// and verifies uplink configuration is restored.
+func teardownVMAgentTest(t *testing.T, data *TestData, vmList []vmInfo) {
+	verifyUpLinkAfterCleanup := func(vm vmInfo) {
+		if err := wait.PollImmediate(30*time.Second, 1*time.Minute, func() (done bool, err error) {
```
the 30 second delay for retry looks pretty long to me. Is that intentional?

Anandkumar26 replied:
Yes. On windows we have observed sometime it may take time. And also earlier, agent was monitoring for ExternalNode every minute.  I can change it to poll for every 10 seconds, with timeout at 1 minute.

antoninbas commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,327 @@
+
+func testExternalNode(t *testing.T, data *TestData, vmList []vmInfo) {
+	verifyExternalNodeRealization := func(vm vmInfo) {
+		if err := wait.PollImmediate(30*time.Second, 1*time.Minute, func() (done bool, err error) {
```
same question as above

Anandkumar26 replied:
Will change the interval duration to 10 seconds.

antoninbas commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,327 @@
+		} else {
+			tempVm = getVMInfo(t, data, vm.nodeName)
+		}
+		assert.Exactly(t, vm.ifName, tempVm.ifName, "Failed to verify uplink interface, Expected %s, Got %s", vm.ifName, tempVm.ifName)
+		assert.Exactly(t, vm.ip, tempVm.ip, "Failed to verify uplink IP, Expected %s, Got %s", vm.ip, tempVm.ip)
```
Are you checking strings? If yes, I am curious as to why you are using `Exactly` instead of `Equal`?

Anandkumar26 replied:
Earlier, I was comparing the vmInfo Object, so I was using Exactly. Will change it to Equal.

antoninbas commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,327 @@
+	var vm vmInfo
+	vm.nodeName = nodeName
+	var cmd string
+	cmd = "ip -o -4 route show to default | awk '{print $5}'"
+	vm.osType = "Linux"
+	rc, ifName, stderr, err := data.RunCommandOnNode(nodeName, cmd)
+	require.NoError(t, err, "Failed to run command <%s> on VM %s, err %v", cmd, nodeName, err)
+	if rc != 0 {
+		t.Errorf("Failed to run command: <%s>, stdout: <%v>, stderr: <%v>", cmd, ifName, stderr)
+	}
+
+	vm.ifName = strings.TrimSpace(ifName)
+	cmd = fmt.Sprintf("ifconfig %s | awk '/inet / {print $2}'| sed 's/addr://'", vm.ifName)
```
as a rule, it's better to do regex matching as part of the test code, rather than expect `awk` and `sed` are available on the test Node

Anandkumar26 replied:
Sure. Noted.. 
Here we had a testbed and then I was trying to get test to work,  So I took this approach.

tnqn commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,327 @@
+// and returns a list of VMs upon success
+func setupVMAgentTest(t *testing.T, data *TestData) ([]vmInfo, error) {
+	t.Logf("List of Windows VMs: '%s', Linux VMs: '%s'", testOptions.windowsVMs, testOptions.linuxVMs)
+	nsReturned, _ := data.clientset.CoreV1().Namespaces().Get(context.TODO(), nameSpace, metav1.GetOptions{})
+	saReturned, _ := data.clientset.CoreV1().ServiceAccounts(nameSpace).Get(context.TODO(), serviceAccount, metav1.GetOptions{})
+	t.Logf("Using service account %s", saReturned.Name)
+	t.Logf("Using namespace %s", nsReturned.Name)
```
why are the 4 lines are needed?
The namespace and serviceAccount are already specified by the constants: `saReturned.Name` must be `nameSpace` and `nsReturned.Name` must be `serviceAccount`.

Anandkumar26 replied:
There are not required. I will just add a single log message for basic troubleshooting purposes using test logs

t.Logf("Using service account %s, namespace %s", serviceAccount, nameSpace)

tnqn commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,327 @@
+
+const (
+	interfaceNameLength = 5
+	nameSpace           = "vm-ns"
```
```suggestion
	namespace           = "vm-ns"
```

Anandkumar26 replied:
Based on previous review comments, I changed it from `namespace` to `nameSpace`

tnqn replied:
which comment? I think namespace is one word, not two?

Anandkumar26 replied:
If you think, it should be `namespace`, I will change it back.

Anandkumar26 replied:
@tnqn: Updated the PR to use namespace.

tnqn commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,327 @@
+func teardownVMAgentTest(t *testing.T, data *TestData, vmList []vmInfo) {
+	verifyUpLinkAfterCleanup := func(vm vmInfo) {
+		if err := wait.PollImmediate(10*time.Second, 1*time.Minute, func() (done bool, err error) {
+			var tempVm vmInfo
```
```suggestion
			var tempVM vmInfo
```

Anandkumar26 replied:
Done

tnqn commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,327 @@
+			} else {
+				tempVm = getWindowsVMInfo(t, data, vm.nodeName)
+			}
+			if !assert.Equal(t, vm.ifName, tempVm.ifName, "Unexpected uplink interface name") {
+				return false, nil
+			}
+			if !assert.Equal(t, vm.ip, tempVm.ip, "Unexpected uplink interface IP") {
+				return false, nil
+			}
```
`!assert.Equal` will mark the test as failure, is it expected? It looks contradictory to mark failure but keep retrying.

Anandkumar26 replied:
We are not running the test with `failfast` mode. So it will retry.
If this marks the test as Failure, then we don't need the assert. 
I will change it to back to comparing two strings.

tnqn commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,327 @@
+				return false, nil
+			}
+			return true, nil
+		}); err != nil {
+			assert.Fail(t, "Failed to verify uplink configuration after cleanup")
+			return
+		}
```
`assert.NoError(t, err, "Failed to verify uplink configuration after cleanup")`

Anandkumar26 replied:
Done

tnqn commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,327 @@
+	for _, vm := range vmList {
+		eeName := generateExternalEntityName(vm)
+		if err := data.crdClient.CrdV1alpha1().ExternalNodes(nameSpace).Delete(context.TODO(), vm.nodeName, metav1.DeleteOptions{}); err != nil {
+			t.Logf("Failed to delete externalnode, err: %v", err)
```
Should it assert NoError for this delete operation?

Anandkumar26 replied:
Yes, it need not assert here. We will try to delete the ExternalNode for the remaining VMs.

tnqn replied:
Unlike `require.NoError`, `assert.NoError` won't terminate the execution but just mark the test as failed. I think it's expected to succeed?
And logs should use upper camel case for K8s resource and CRD kind names, i.e.  ExternalNode. There are few other places too, e.g. L73

tnqn commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,327 @@
+			exists, err := verifyInterfaceIsInOVS(t, data, vm)
+			return exists, err
+		}); err != nil {
+			require.NoError(t, err, "Failed to verify host interface in OVS, vmInfo %+v", vm)
```
Using `NoError` doesn't need an extra `err != nil` check

Anandkumar26 replied:
Done

tnqn commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,327 @@
+			return
+		}
+
+		var tempVm vmInfo
```
```suggestion
		var tempVM vmInfo
```

Anandkumar26 replied:
Done

tnqn commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,327 @@
+		} else {
+			tempVm = getVMInfo(t, data, vm.nodeName)
+		}
+		assert.Equal(t, vm.ifName, tempVm.ifName, "Failed to verify uplink interface, Expected %s, Got %s", vm.ifName, tempVm.ifName)
+		assert.Equal(t, vm.ip, tempVm.ip, "Failed to verify uplink IP, Expected %s, Got %s", vm.ip, tempVm.ip)
```
`assert.Equal` will dump expected value and actual value when the check fails, so you can remove "Expected %s, Got %s" from the message.

Anandkumar26 replied:
Done

tnqn commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,327 @@
+	rc, ifName, stderr, err := data.RunCommandOnNode(nodeName, cmd)
+	require.NoError(t, err, "Failed to run command <%s> on VM %s, err %v", cmd, nodeName, err)
+	if rc != 0 {
+		t.Errorf("Failed to run command: <%s>, stdout: <%v>, stderr: <%v>", cmd, ifName, stderr)
```
By using `t.Errorf`, the following code will continue running, which will lead to unexpected result. I think it should use `require.Equal(t, 0, rc, "... stdout: <%v> ...", cmd, ifName)` to stop execution on mismatch

Anandkumar26 replied:
Done.

tnqn commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,327 @@
+	rc, ifIP, stderr, err := data.RunCommandOnNode(nodeName, cmd)
+	require.NoError(t, err, "Failed to run command <%s> on VM %s, err %v", cmd, nodeName, err)
+	if rc != 0 {
+		t.Errorf("Failed to run command: <%s>, stdout: <%v>, stderr: <%v>", cmd, ifIP, stderr)
```
ditto

Anandkumar26 replied:
Done

tnqn commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,327 @@
+	rc, ifIndex, stderr, err := data.RunCommandOnNode(nodeName, cmd)
+	require.NoError(t, err, "Failed to run command <%s> on VM %s, err %v", cmd, nodeName, err)
+	if rc != 0 {
+		t.Errorf("Failed to run command: <%s>, stdout: <%v>, stderr: <%v>", cmd, ifIndex, stderr)
```
ditto

Anandkumar26 replied:
Done

tnqn commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,327 @@
+	if err != nil {
+		return false, fmt.Errorf("failed to run command <%s> on VM %s, err %v", cmd, vm.nodeName, err)
+	}
+	if rc != 0 {
+		return false, fmt.Errorf("failed to run command: <%s>, stdout: <%v>, stderr: <%v>", cmd, stdout, stderr)
+	}
+
+	if strings.Contains(stdout, "no row") {
+		return false, fmt.Errorf("failed to find OVS port %s on VM %s, err %v", vm.ifName, vm.nodeName, err)
+	}
+
+	if strings.Contains(stdout, vm.ifName) && strings.Contains(stdout, "name") {
+		return true, nil
+	}
```
when `ovs-vsctl --column=name list port` specifies port name, I think `rc` already represent whether the port exists? The following two checks seems useless

Anandkumar26 replied:
We are interested only in the below outputs and it should retry for any other  output (irrespective of status code).

``` 
ovs-vsctl: no row "ens192" in table Port --> failure case
name                : eth0 --> success case
```


So I'm changing this into.
```
	rc, stdout, stderr, err := data.RunCommandOnNode(vm.nodeName, cmd)
	if err != nil {
		return false, fmt.Errorf("failed to run command <%s> on VM %s, err %v", cmd, vm.nodeName, err)
	}

	if strings.Contains(stdout, "no row") {
		return false, fmt.Errorf("failed to find OVS port %s on VM %s, err %v, rc %d, stdout %v, stderr %v", vm.ifName, vm.nodeName, err, rc, stdout, stderr)
	}

	if strings.Contains(stdout, vm.ifName) && strings.Contains(stdout, "name") {
		return true, nil
	}
```

tnqn commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,327 @@
+	testEn.SetName(nameSpace, nodeName)
+	var ipList []string
+	ipList = append(ipList, ip)
+	testEn.Spec.Interfaces = append(testEn.Spec.Interfaces, crdv1alpha1.NetworkInterface{
+		Name: ifName,
+		IPs:  ipList,
+	})
```
I think the point of having `ExternalNodeSpecBuilder` is to encapsulate such code, otherwise it looks inconsistent to use both builder and raw code to construct an object. The builder should have a `AddInterface(name string, ips []string)` method.

Anandkumar26 replied:
Sure.

tnqn commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,327 @@
+	return data.crdClient.CrdV1alpha1().ExternalNodes(nameSpace).Create(context.TODO(), testEn.Get(), metav1.CreateOptions{})
+}
+
+func generateExternalEntityName(vm vmInfo) (eeName string) {
```
I think there is a public `GenExternalEntityName` function?

Anandkumar26 replied:
yes, I just saw it. I will update the PR

Anandkumar26 commented on ci/jenkins/test-vm.sh:
```diff
@@ -0,0 +1,264 @@
+        fi
+    fi
+    echo "====== Applying Antrea yaml ======"
+    ./hack/generate-manifest.sh --export-controller-nodeport 32767 --enable-externalnode > ${WORKDIR}/antrea.yml
```
I will update it and try it out.

Anandkumar26 commented on hack/generate-manifest.sh:
```diff
@@ -184,6 +188,14 @@ case $key in
     HELM_VALUES+=("$2")
     shift 2
     ;;
+    --export-controller-nodeport)
+    ANTREA_SERVICE_NODEPORT=$2
+    shift 2
+    ;;
+    --enable-externalnode)
+    ENABLE_EXTERNALNODE=true
+    shift
+    ;;
```
Will update it based on your suggestion.

Anandkumar26 commented:
/test-vm-e2e

Anandkumar26 commented:
/test-vm-e2e

tnqn commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,297 @@
+		vms := strings.Split(testOptions.linuxVMs, ",")
+		for _, vm := range vms {
+			t.Logf("Get info for Linux VM: %s", vm)
+			tempVm := getVMInfo(t, data, vm)
```
```suggestion
			tempVM := getVMInfo(t, data, vm)
```
there are few other places too

Anandkumar26 replied:
I thought updated it. but missed 2 instances.

tnqn commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,297 @@
+				tempVM = getWindowsVMInfo(t, data, vm.nodeName)
+			}
+			if vm.ifName != tempVM.ifName {
+				t.Logf("Retry, unexpected uplink interface name, Expected %s, Got %s", vm.ifName, tempVM.ifName)
```
```suggestion
				t.Logf("Retry, unexpected uplink interface name, expected %s, got %s", vm.ifName, tempVM.ifName)
```

Anandkumar26 replied:
Done

tnqn commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,297 @@
+			exists, err := verifyInterfaceIsInOVS(t, data, vm)
+			return exists, err
+		})
+		{
```
why keeping this brace?

Anandkumar26 replied:
Addressed it.

tnqn commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,297 @@
+	}
+
+	if strings.Contains(stdout, "no row") {
+		return false, fmt.Errorf("failed to find OVS port %s on VM %s, err %v, rc %d, stdout %v, stderr %v", vm.ifName, vm.nodeName, err, rc, stdout, stderr)
```
Do we expect the OVS port to be found immediately after the test starts? returnning error will terminate `PollImmediate` and return error, is it expected?

Anandkumar26 replied:
Actually we dont want to return error and abort PollImmediate. 
But we do want to have special handling for this output. let me change return nil and just the log the output.

wenyingd commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,297 @@
+			return
+		}
+
+		var tempVM vmInfo
```
line 179-186 is not reachable.

Anandkumar26 replied:
Good catch. Will change it as below.
```
		err := wait.PollImmediate(10*time.Second, 1*time.Minute, func() (done bool, err error) {
			t.Logf("Verify host interface configuration for VM: %s", vm.nodeName)
			exists, err := verifyInterfaceIsInOVS(t, data, vm)
			return exists, err
		})
		assert.NoError(t, err, "Failed to verify host interface in OVS, vmInfo %+v", vm)
```

Anandkumar26 commented:
/test-vm-e2e

Anandkumar26 commented:
/test-vm-e2e

Anandkumar26 commented:
Addressed the review comments. @tnqn @wenyingd please take a look at the latest patchset. 
If there are any no other comments, then we could merge the PR>

The test is passing with latest patchset.
http://10.176.27.169:8080/job/antrea-vm-e2e-for-pull-request/130

tnqn commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,295 @@
+// and returns a list of VMs upon success
+func setupVMAgentTest(t *testing.T, data *TestData) ([]vmInfo, error) {
+	t.Logf("List of Windows VMs: '%s', Linux VMs: '%s'", testOptions.windowsVMs, testOptions.linuxVMs)
+	t.Logf("Using ServiceAccount %s, NameSpace %s", serviceAccount, namespace)
```
Namespace

Anandkumar26 replied:
Done.

tnqn commented on test/e2e/vmagent_test.go:
```diff
@@ -0,0 +1,295 @@
+	t.Logf("TestVMAgent teardown")
+	for _, vm := range vmList {
+		if err := data.crdClient.CrdV1alpha1().ExternalNodes(namespace).Delete(context.TODO(), vm.nodeName, metav1.DeleteOptions{}); err != nil {
+			t.Logf("Failed to delete ExternalNode, err: %v", err)
```
what about this? https://github.com/antrea-io/antrea/pull/3824#discussion_r941208097

Anandkumar26 replied:
synced offline and updated the PR.

Anandkumar26 commented:
/test-vm-e2e

tnqn reviewed (approved):
LGTM

Anandkumar26 commented:
Removed unused variable.

Anandkumar26 commented:
/test-vm-e2e
